driftbase/counters.py

```diff
@@ -54,60 +54,71 @@ def get_counter(counter_key):
-        try:
-            row = Counter(name=name, counter_type=counter_type)
-            db_session.add(row)
-            db_session.commit()
-        except IntegrityError as e:
-            # if someone has inserted the counter in the meantime, retrieve it
-            if "duplicate key" in repr(e):
-                db_session.rollback()
-                row = db_session.query(Counter).filter(Counter.name == name).first()
-            else:
-                raise
 
-        clear_counter_cache()
-    counter_id = row.counter_id
+def batch_create_player_counters(player_id, counter_ids, db_session=None):
+    """
+    Create all missing player counters for the specified counter IDs
+    """
+    if not db_session:
+        db_session = g.db
 
-    return counter_id
+    values = [{"counter_id": counter_id, "player_id": player_id} for counter_id in counter_ids]
+    insert_clause = insert(PlayerCounter).values(values)
+    fallback_clause = insert_clause.on_conflict_do_nothing(index_elements=['counter_id', 'player_id'])
+    return db_session.execute(fallback_clause)
 
 
-def get_or_create_player_counter(counter_id, player_id):
-    player_counter = g.db.query(PlayerCounter) \
-        .filter(PlayerCounter.player_id == player_id,
-                PlayerCounter.counter_id == counter_id) \
-        .first()
+def batch_update_counter_entries(player_id, entries, db_session=None):
+    if not db_session:
+        db_session = g.db
+
+    absolute_values = []
+    counter_values = []
+    for k, e in entries.items():
+        for period in COUNTER_PERIODS:
+            date_time = get_date_time_for_period(period, e["timestamp"])
+            entry = dict(counter_id=e["counter_id"], player_id=player_id, period=period, date_time=date_time,
+                         value=e["value"])
+            if e["is_absolute"]:
+                absolute_values.append(entry)
+            else:
+                counter_values.append(entry)
+
+    if len(absolute_values):
+        insert_clause = insert(CounterEntry).values(absolute_values)
+        update_clause = insert_clause.on_conflict_do_update(
+            index_elements=['counter_id', 'player_id', 'period', 'date_time'],
+            set_=dict(value=insert_clause.excluded.value))
+        db_session.execute(update_clause)
+        db_session.commit()
 
-    if not player_counter:
-        log.info("Creating new player counter for counter_id %s", counter_id)
-        player_counter = PlayerCounter(counter_id=counter_id, player_id=player_id)
-        g.db.add(player_counter)
-    g.db.commit()
-    return player_counter
+    if len(counter_values):
+        insert_clause = insert(CounterEntry).values(counter_values)
+        update_clause = insert_clause.on_conflict_do_update(
+            index_elements=['counter_id', 'player_id', 'period', 'date_time'],
+            set_=dict(value=CounterEntry.value + insert_clause.excluded.value))
+        db_session.execute(update_clause)
+        db_session.commit()
```

Comment:
Is there a reason we commit() inside both if-blocks rather than once at the end?

Reply:
Not anymore, there was an issue with the query when it had a different format, but it should be fined to commit both at once

Reply:
Potentially there's a scenario where we'd want to split the update into multiple statements due to size, but I need to find if there's such a limit first